### PR TITLE
Lighten communication between peer and tracker

### PR DESCRIPTION
**Pull request by minsulee2**
Updates:
- No more web socket between peers and tracker
- Add POST /update_peer_info
- Update Genesis params

### CODE REVIEW

File: blockchain-configs/afan-shard/blockchain_params.json [hunk 1]
```diff
@@ -25,7 +25,7 @@
     "NETWORK_ID": 0,
     "CHAIN_ID": 0,
     "HOSTING_ENV": "local",
-    "TRACKER_WS_ADDR": "ws://localhost:6000",
+    "TRACKER_ENDPOINT_URL": "http://localhost",
```
Comment (platfowner) [inline blockchain-configs/afan-shard/blockchain_params.json:28]: TRACKER_UPDATE_PEER_INFO_URL?

File: tracker-server/index.js [hunk 2]
```diff
@@ -67,14 +60,23 @@ app.get('/network_status', (req, res, next) => {
...
+  res.render(__dirname + '/index.html', {}, (err, html) => {
     const networkStatus = getNetworkStatus();
-    const graphData = await getGraphData(networkStatus);
+    const graphData = getGraphData(networkStatus);
     html = html.replace(/{ \/\* replace this \*\/ };/g, JSON.stringify(graphData));
     res.send(html);
   });
 });
 
+app.post('/update_peer_info', (req, res) => {
```
Comment (platfowner) [inline tracker-server/index.js:71]: sorry for this late feedback, but i just realized that we already have json rpc interface for the tracker server:
https://github.com/ainblockchain/ain-blockchain/blob/develop/tracker-server/json-rpc.js

what do you think about using the json rpc api instead?

File: tracker-server/util.js [hunk 3]
```diff
@@ -1,7 +1,14 @@
+const PEER_LIVENESS_THRESHOLD_MS = 5 * 60 * 1000   // 5 minutes
```
Comment (platfowner) [inline tracker-server/util.js:1]: what about increasing TRACKER_UPDATE_INTERVAL_MS as well, e.g. to 2.5 minutes?

File: tracker-server/index.js [hunk 4]
```diff
@@ -88,122 +93,18 @@ process.on('uncaughtException', function(err) {
...
-    `  memory: ${memoryFreeMb}MB,\n` +
-    `  peers:\n` +
-    `    outbound (${_.get(nodeInfo, 'networkStatus.connectionStatus.outgoingPeers')}),\n` +
-    `    inbound (${_.get(nodeInfo, 'networkStatus.connectionStatus.incomingPeers')}),\n` +
-    `  updatedAt: ${nodeInfo.updatedAt}`;
+  return Object.values(peerNodes).filter(info => isPeerAlive(info.updatedAt)).length;
 }
 
-function getNodeLocation(ip) {
+function getPeerLocation(ip) {
```
Comment (platfowner) [inline tracker-server/index.js:107]: what about using 'Node' as they're not peers from the tracker's perspective?

File: tracker-server/index.js [hunk 5]
```diff
@@ -88,122 +93,18 @@ process.on('uncaughtException', function(err) {
...
-  return `[${abbrAddr(nodeInfo.address)} (${ip})]:\n` +
-    `  isAlive: ${nodeInfo.isAlive},\n` +
-    `  state: ${_.get(nodeInfo, 'nodeStatus.state')},\n` +
-    `  disk: ${diskAvailableMb}MB,\n` +
-    `  memory: ${memoryFreeMb}MB,\n` +
-    `  peers:\n` +
-    `    outbound (${_.get(nodeInfo, 'networkStatus.connectionStatus.outgoingPeers')}),\n` +
-    `    inbound (${_.get(nodeInfo, 'networkStatus.connectionStatus.incomingPeers')}),\n` +
-    `  updatedAt: ${nodeInfo.updatedAt}`;
+  return Object.values(peerNodes).filter(info => isPeerAlive(info.updatedAt)).length;
```
Comment (platfowner) [inline tracker-server/index.js:104]: what about 'isNodeAlive' from the tracker's perspective and passing by 'info' itself?

File: tracker-server/index.js (hunk 2 above)
Comment (platfowner) [reply tracker-server/index.js:74]: and peer -> node?

File: blockchain-configs/base/blockchain_params.json [hunk 6]
```diff
@@ -24,7 +24,7 @@
     "NETWORK_ID": 0,
     "CHAIN_ID": 0,
     "HOSTING_ENV": "local",
-    "TRACKER_WS_ADDR": "ws://localhost:5000",
+    "TRACKER_UPDATE_PEER_INFO_URL": "http://localhost:8080/update_peer_info",
```
Comment (platfowner) [inline blockchain-configs/base/blockchain_params.json:27]: what about 'TRACKER_UPDATE_NODE_INFO_URL' or just 'TRACKER_UPDATE_URL'?

File: p2p/index.js [hunk 7]
```diff
@@ -170,29 +168,24 @@ class P2pClient {
...
-
-  setIntervalForTrackerConnection() {
-    if (!this.intervalTrackerConnection) {
-      this.intervalTrackerConnection = setInterval(() => {
-        this.connectToTracker();
-      }, TRACKER_RECONNECTION_INTERVAL_MS);
+  /**
+   * Update peer info to tracker via POST.
+   */
+  async updatePeerInfoToTracker() {
```
Comment (platfowner) [inline p2p/index.js:174]: similarly what about 'updateNodeInfoToTracker'?

File: blockchain-configs/afan-shard/blockchain_params.json (hunk 1 above)
Comment (liayoo) [reply blockchain-configs/afan-shard/blockchain_params.json:28]: \+ it's missing `:8080/update_peer_info`?

File: tracker-server/util.js (hunk 3 above)
Comment (platfowner) [reply tracker-server/util.js:1]: I think a todo would be good at this moment as we're doing optimization intensively these days.

File: p2p/index.js (hunk 7 above)
Comment (minsulee2) [reply p2p/index.js:174]: Updated!

File: blockchain-configs/base/blockchain_params.json (hunk 6 above)
Comment (minsulee2) [reply blockchain-configs/base/blockchain_params.json:27]: I took `TRACKER_UPDATE_URL` Thanks!

File: tracker-server/index.js (hunk 4 above)
Comment (minsulee2) [reply tracker-server/index.js:107]: Fare enough. Thanks!

File: tracker-server/util.js (hunk 3 above)
Comment (minsulee2) [reply tracker-server/util.js:1]: Left NOTE for this! Thanks!

File: blockchain-configs/afan-shard/blockchain_params.json (hunk 1 above)
Comment (minsulee2) [reply blockchain-configs/afan-shard/blockchain_params.json:28]: Done!

File: tracker-server/index.js (hunk 5 above)
Comment (minsulee2) [reply tracker-server/index.js]: Updated!

File: tracker-server/index.js (hunk 2 above)
Comment (minsulee2) [reply tracker-server/index.js]: It took time a bit tho! Updated! Now, node info is updated via json-rpc!

**Comment by minsulee2** [comment]
PTAL! All addressed!

File: tracker-server/tracker.js [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+  abbrAddr,
+  isNodeAlive
+} = require('./util');
+const { BlockchainConfigs } = require('../common/constants');
+
+const DISK_USAGE_PATH = os.platform() === 'win32' ? 'c:' : '/';
+
+class Tracker {
+  constructor() {
+    this.peerNodes = {};
```
Comment (platfowner) [inline tracker-server/tracker.js:17]: what about 'blockchainNodes' as the tracker's main role is not to suggest peers any more?

File: tracker-server/tracker.js [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+  }
+
+  setPeerNodes(peerInfo) {
```
Comment (platfowner) [inline tracker-server/tracker.js:20]: s/Peer/Blockchain/?

File: tracker-server/json-rpc.js [hunk 10]
```diff
@@ -7,33 +7,40 @@
  * @return {dict} A closure of functions compatible with the jayson library for
  *                  servicing JSON-RPC requests.
  */
-module.exports = function getMethods(nodes) {
+module.exports = function getMethods(tracker) {
+  const peerNodes = tracker.peerNodes;
```
Comment (platfowner) [inline tracker-server/json-rpc.js:11]: s/peerNodes/blockchainNodes/?

File: tracker-server/tracker.js (hunk 8 above)
Comment (minsulee2) [reply tracker-server/tracker.js]: Indeed! Updated!

File: tracker-server/tracker.js (hunk 9 above)
Comment (minsulee2) [reply tracker-server/tracker.js]: Updated!

File: tracker-server/json-rpc.js (hunk 10 above)
Comment (minsulee2) [reply tracker-server/json-rpc.js]: blockchainNode!

**Comment by minsulee2** [comment]
PTAL!

File: tracker-server/tracker.js [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+  abbrAddr,
+  isNodeAlive
+} = require('./util');
+const { BlockchainConfigs } = require('../common/constants');
+
+const DISK_USAGE_PATH = os.platform() === 'win32' ? 'c:' : '/';
+
+class Tracker {
+  constructor() {
+    this.blockchainNode = {};
```
Comment (platfowner) [inline tracker-server/tracker.js:17]: what about plural form (blockchainNodes)?

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks!

File: start_node_genesis_gcp.sh [hunk 12]
```diff
@@ -133,12 +133,12 @@ fi
 KEYSTORE_DIR=testnet_dev_staging_keys
 if [[ $SEASON = 'spring' ]]; then
     export BLOCKCHAIN_CONFIGS_DIR=blockchain-configs/testnet-prod
-    export TRACKER_WS_ADDR=ws://35.221.137.80:5000
+    export TRACKER_UPDATE_URL=http://35.221.137.80:8080/json-rpc
```
Comment (liayoo) [inline start_node_genesis_gcp.sh:136]: we could explicitly say it's a json rpc URL as well? (e.g. `TRACKER_UPDATE_JSON_RPC_URL`)

**Review by liayoo — APPROVED** [review]
LGTM with a comment. Thanks!

File: tracker-server/tracker.js (hunk 11 above)
Comment (minsulee2) [reply tracker-server/tracker.js]: Done!

File: start_node_genesis_gcp.sh (hunk 12 above)
Comment (minsulee2) [reply start_node_genesis_gcp.sh]: Done!

**Comment by minsulee2** [comment]
Thanks for the reviews! It took time a bit bcuz of integration test! Merging now!

**Pull request merged**